### PR TITLE
Update Pack3d to accept older input file formats

### PR DESCRIPTION
**Pull request by ted-authentise**
### Pack3d Updates 2024

These changes allow the current version of pack3d to use input files generated for the `master-ricoh` version of pack3d.

### Changes
- Handle missing `axes_lock` value in input json. This allows current version to use input files generated for the `master-ricoh` version of pack3d
- Automated tests (basic "will it run tests" for current and old input formats, plus some rotation locking validation) + tests in CI
- Tidy up, moving code out of the main function, separating out config parsing and the main algorithm loop.
- Add `save_stl` flag to output generated layout as stl (previously only possible by uncommenting code)

### Go Primer

The `/cmd` directory contains the location of executable entrypoints in a go project. The only one we use for pack3d is `pack3d`. The `main` function is the entrypoint. The algorithm you'll find in `pack3d/pack3d.go` is directly lifted from the `main` function; I've moved this into a package for better readability.

### CODE REVIEW

File: pack3d/pack3d_test.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+func TestCoPack(t *testing.T) {
+	config, err := pack3d.ParseConfig("../tests/fixtures/copack.json")
+
+	fmt.Printf("%v", config)
+	if err != nil {
+		t.Fatalf("Failed to load config: %s", err)
+	}
+
+	_, err = pack3d.Pack(config)
```
Comment (francis-authentise) [inline pack3d/pack3d_test.go:18]: anyway we could assert/validate the packing output here, if it matters?

File: pack3d/pack3d.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
+package pack3d
```
Comment (francis-authentise) [inline pack3d/pack3d.go:1]: Cool! I've always wanted to see how the packing algorithm worked

**Review by francis-authentise — APPROVED** [review]
Refactor looks good, one small remark below

File: pack3d/pack3d.go [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+			// 6. coarse approx of its volume.
+			totalVolume += mesh.BoundingBox().Volume()
+
+		} else {
+
+			coPackMap[item.Filename] = item.Copack
+
+			// 1a. load the main co-packing mesh (the "parent" co-packing mesh, so to say).
+			done = timed(fmt.Sprintf("loading the main co-packing mesh %s", item.Filename))
+			mesh, err = fauxgl.LoadMesh(item.Filename)
```
Comment (hardypete-etp) [inline pack3d/pack3d.go:116]: Have only just realised that Fogelman called this library as a pun on their name...

File: pack3d/utils_test.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+	mtx := config.ManufacturingOrientation()
+
+
+	if mtx != fauxgl.Identity() {
+		t.Fatal("Unexpected manufacturing transformation matrix for unlocked axes")
+	}
+
+	// Legacy master-ricoh case, axes lock was not supplied
+	config.AxesLock = nil
```
Comment (hardypete-etp) [inline pack3d/utils_test.go:77]: 💯

**Review by hardypete-etp — APPROVED** [review]
Sensible refactor and excellent base test coverage on the two merged use cases 👍 
Love the test integration into github actions!

File: pack3d/pack3d_test.go (hunk 1 above)
Comment (ted-authentise) [reply pack3d/pack3d_test.go:18]: Added this in the other PR!

File: pack3d/pack3d.go (hunk 2 above)
Comment (ted-authentise) [reply pack3d/pack3d.go:1]: Yeah it's an interesting one and definitely room for improvement. If you want a deeper dive, learning about the BVH trees (representation we build for quick collision detection) and annealing (gradual improvement in packing efficiency) are great places to start.

File: pack3d/pack3d.go (hunk 3 above)
Comment (ted-authentise) [reply pack3d/pack3d.go:116]: ...same here...

**Comment by far-authentise** [comment]
@ted-authentise this one PR makes this program like 30% better in one shot.

**Pull request merged**